File: src/main/java/ru/ioleksiv/telegram/bot/core/model/method/Action.java
package ru.ioleksiv.telegram.bot.core.model.method;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import ru.ioleksiv.telegram.bot.core.model.responses.CommonResponse;

import java.util.Optional;

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
public abstract class Action<RES> {

    @JsonIgnore
    public abstract Optional<RES> send();

    @JsonIgnore
    protected abstract Class<? extends CommonResponse<RES>> getResultWrapperClass();
}

File: server-api/src/main/kotlin/io/github/oleksivio/tl/kbot/server/api/model/method/Action.kt
package io.github.oleksivio.tl.kbot.server.api.model.method

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import io.github.oleksivio.tl.kbot.server.api.model.CommonResponse
import kotlin.reflect.KClass

@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
abstract class Action<RES> {

    abstract val method: String

    abstract val resultWrapperClass: KClass<out CommonResponse<RES>>

}
